src/kudu/consensus/log_cache.cc

```diff
@@ -164,17 +197,182 @@ void LogCache::TruncateOpsAfterUnlocked(int64_t index) {
+  VLOG(2) << "Uncompressing messages before writing to log." <<
+             " opid: " << msg->get()->id().ShortDebugString() <<
+             " codec: " << compression_codec <<
+             " op_type: " << op_type <<
+             " compressed payload size: " << compressed_size <<
+             " uncompressed payload size: " << uncompressed_size;
+
+  // Resize buffer to hold uncompressed payload.
+  // TODO: needs perf testing and maybe implement streaming (un)compression
+  buffer.resize(uncompressed_size);
+  Slice compressed_slice(payload.payload().c_str(), compressed_size);
+
+  Status status;
+  const CompressionCodec* codec = codec_.load();
+  if (codec && codec->type() == compression_codec) {
+    status = codec->Uncompress(
+        compressed_slice, buffer.data(), uncompressed_size);
+  }
+  else {
+    // Either compression is not enabled on this instance OR this message uses a
+    // different compression codec. Get the right codec now
+    status = GetCompressionCodec(compression_codec, &codec);
+    if (status.ok()) {
+      status = codec->Uncompress(
+          compressed_slice, buffer.data(), uncompressed_size);
+    }
+  }
+
+  // Return early if uncompression failed
+  RETURN_NOT_OK_PREPEND(status,
+      Substitute("Failed to uncompress OpId $0. Compression codec used: $1, "
+                 "Operation type: $2, Compressed payload size: $3"
+                 "Uncompressed payload size: $4",
+                 msg->get()->id().ShortDebugString(), compression_codec,
+                 op_type, compressed_size, uncompressed_size));
+
+  // Now create a new ReplicateMsg and copy over the contents from the original
+  // msg and the uncompressed payload
+  std::unique_ptr<ReplicateMsg> rep_msg(new ReplicateMsg);
+  *(rep_msg->mutable_id()) = msg->get()->id();
+  rep_msg->set_timestamp(msg->get()->timestamp());
+  rep_msg->set_op_type(msg->get()->op_type());
+
+  WritePayloadPB* write_payload = rep_msg->mutable_write_payload();
+  write_payload->set_payload(std::move(buffer.ToString()));
+
+  uncompressed_msg->reset(rep_msg.release());
+  return Status::OK();
+}
+
+Status LogCache::CompressMsg(const ReplicateRefPtr& msg,
+                             faststring& buffer,
+                             std::unique_ptr<ReplicateMsg>* compressed_msg) {
+  // Grab the reference to the payload that needs to be compressed
+  ReplicateMsg* original_msg = msg->get();
+  const std::string& payload_str = original_msg->write_payload().payload();
+  bool is_compressed =
+    (original_msg->write_payload().compression_codec() != NO_COMPRESSION);
+
+  if (PREDICT_FALSE(is_compressed)) {
+    // msg is already compressed
+    return Status::NotSupported("Double compression is not supported");
+  }
+
+  if (original_msg->op_type() != WRITE_OP_EXT) {
+    // Compression of messages other than WRITE_OP_EXT is not supported
+    return Status::NotSupported("Only write ops can be compressed");
+  }
+
+  const CompressionCodec* codec = codec_.load();
```

Comment:
nit: we probably should be able to use something lesser than seq_cst because CompressionCodec
can be made to change safely in the ring by triggering a promotion

Reply:
Will stick with this now and avoid such micro optimization